source/common/upstream/upstream_impl.cc

```diff
@@ -262,12 +262,34 @@ ResourceManager& ClusterInfoImpl::resourceManager(ResourcePriority priority) con
   return *resource_managers_.managers_[enumToInt(priority)];
 }
 
-void ClusterImplBase::blockHcUpdates(bool block) {
-  ASSERT(block_hc_updates_ == !block);
-  block_hc_updates_ = block;
+void ClusterImplBase::setInitialized() {
+  if (!initialized_ && health_checker_ && pending_initialize_health_checks_ == 0) {
+    pending_initialize_health_checks_ = hosts().size();
+    ASSERT(pending_initialize_health_checks_ > 0);
+
+    // TODO(mattklein123): Remove this callback when done.
+    health_checker_->addHostCheckCompleteCb([this](HostSharedPtr, bool) -> void {
+      if (pending_initialize_health_checks_ > 0 && --pending_initialize_health_checks_ == 0) {
+        setInitialized();
+      }
+    });
+  }
 
-  if (!block_hc_updates_) {
-    reloadHealthyHosts();
+  initialized_ = true;
+  if (initialize_callback_ && pending_initialize_health_checks_ == 0) {
+    // Every time a host changes HC state we cause a full healthy host recalculation which
+    // for expensive LBs (ring, subset, etc.) can be quite time consuming. During startup, this
+    // can also block worker threads by doing this repeatedly. There is no reason to do this
+    // as we will not start taking traffic until we are initialized. By blocking HC updates
```

Comment:
It's a bit weird to have this comment here, where we're not so much blocking as making sure we update. 

Do you think it makes sense to move this to where we block the reload, and comment here "we're doing a reload because we've been blocking them [see comment over there"?

Reply:
sure